File: algorithms/hard/vertical_order_traversal_of_a_binary_tree.ex
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec vertical_traversal(root :: TreeNode.t() | nil) :: [[integer]]
  def vertical_traversal(root) do
    map = aux(%{}, root, 0, 0)

    map
    |> Map.keys()
    |> Enum.sort()
    |> Enum.map(fn k ->
      map[k]
      |> Enum.sort()
      |> Enum.map(fn {_r, v} -> v end)
    end)
  end

  def aux(map, nil, _r, _c), do: map

  def aux(map, root, r, c) do
    map
    |> Map.update(c, [{r, root.val}], &[{r, root.val} | &1])
    |> aux(root.left, r + 1, c - 1)
    |> aux(root.right, r + 1, c + 1)
  end
end

File: algorithms/hard/vertical_order_traversal_of_a_binary_tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::collections::BTreeMap;
use std::rc::Rc;

fn dfs(
    root: Option<Rc<RefCell<TreeNode>>>,
    map: &mut BTreeMap<i32, Vec<(i32, i32)>>,
    r: i32,
    c: i32,
) {
    if let Some(node) = root {
        map.entry(c).or_insert(vec![]).push((r, node.borrow().val));
        dfs(node.borrow().left.clone(), map, r + 1, c - 1);
        dfs(node.borrow().right.clone(), map, r + 1, c + 1);
    }
}

impl Solution {
    pub fn vertical_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut map = BTreeMap::new();
        dfs(root, &mut map, 0, 0);
        let mut ans = Vec::new();
        for mut arr in map.into_values() {
            arr.sort();
            ans.push(arr.into_iter().map(|(_, v)| v).collect());
        }
        ans
    }
}
